### PR TITLE
fix(textarea): textarea with autogrow will size to its contents

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Resize behavior currently is using the scroll height of the control, to resize the outer container. This results in flaky behavior where the textarea will not correctly grow to its contents.

Issue Number: #24793, #21242

## What is the new behavior?

- Textarea sizing is controlled through a pseudo element with the same visual appearance as the textarea (only hidden)
- Pseudo element reflows naturally, without manually assigning container height: (inspired from: https://css-tricks.com/the-cleanest-trick-for-autogrowing-textareas/).
- Re-introduces screenshot tests for auto-grow behavior

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by Lxstr** [comment]
Hi @sean-perkins.  I would really like to try this solution but I am having trouble using a forked @ionic/core in my stenciljs project. I did manage a while ago as I was having trouble with textarea [https://github.com/ionic-team/ionic-framework/issues/21242](url) but for the life of me can't figure out how I did it. I have looked at the contributing guide but it's vague. And I've tried npm link but got confused.

So far I have 

- forked ionic-framework
- cloned to laptop
- merged feat/textarea-auto-grow to main
- $ cd /core
- $ npm run build
- removed anything in .gitignore starting with core/
- commited the build
- pushed to remote (https://github.com/Lxstr/ionic-framework.git)

From here how would I use this build in a way that I could easily swap back and forth with the real ionic release? 

Ideally have both installed in my project and just import the one I want to test in the appropriate component.

Thanks in advance

**Comment by Lxstr** [comment]
I've tested the new textarea and the bugs I was previously experiencing seem to be solved!! Awesome!

However, I do notice issues with the padding. It seems like the ::after elements is creating so much padding that it seems like an entire extra line is added at the bottom of the textarea that is un-useable.

I think the expectation would be little to no padding that the user can then add if required on textareas?

Thanks again for the fix and looking forward to the merge!

Screen Shot 2022-01-10 at 5 20 55 pm
Screen Shot 2022-01-10 at 5 34 14 pm

**Comment by sean-perkins** [comment]
This PR is on hold until we finish our Playwright integration (being able to test across different browsers). Will rebase/resolve merge conflicts at that time.

**Comment by anagstef** [comment]
Hello everyone! This PR indeed fixes all the issues we have with textarea and autogrow. Thank you for resolving them!!!

Is this getting prioritised anytime soon? I see that some Playwright PRs are merged. 

Would love to help if I can. Thank you again! :)

**Comment by sean-perkins** [comment]
Since Playwright was merged in with our 6.1.0 work, this PR is ready to be worked on and updated for 6.2.0. I cannot commit to a timeline on 6.2, since we just released, but you can expect this feature to be included in those changes.

File: core/src/components/textarea/test/basic/index.html
Comment (amandaejohnston) [inline core/src/components/textarea/test/basic/index.html:74]: Is this needed? Seems like the `autogrow` test covers it, and that'd be one less thing to worry about in the `basic` test.

Comment (sean-perkins) [reply core/src/components/textarea/test/basic/index.html]: I think we can remove it. It doesn't add any test coverage that isn't already handled by the existing auto grow tests.

**Review by amandaejohnston — APPROVED** [review]

File: core/src/components/textarea/textarea.scss
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:86]: Can we add comments here as to why this technique works?

File: core/src/components/textarea/textarea.scss
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:126]: Does this fix the scrollbar showing when zooming in/out?

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

**Comment by liamdebeasi** [comment]
There is also some strange flickering on iOS 13:
https://user-images.githubusercontent.com/2721089/179781340-6312fdb3-2cea-4095-9f6b-7f1b0097d729.mov

File: core/src/components/textarea/textarea.scss
Comment (sean-perkins) [reply core/src/components/textarea/textarea.scss:124]: This is to resolve the vertical scrollbar showing on Firefox. It may resolve other issues with the scrollbar as well though.

**Comment by sean-perkins** [comment]
@liamdebeasi good find. Looks like iOS 13 has some rendering difficulties with `grid-area` with multiple elements in a single container. 

To work around this, the reflected pseudo element will be the only element affecting the layout. The native textarea will be absolutely positioned and fill the container's height when the textarea is using auto grow. 

This appears to resolve the iOS 13 rendering bug and maintains auto grow behavior.

**Review by liamdebeasi — COMMENTED** [review]
It looks like this may actually be a scroll assist issue with duplicating the input while we adjust the scroll position

https://user-images.githubusercontent.com/2721089/181077447-e500effe-ea17-40f9-b9f2-ac5cb4e841fb.mp4

Setting the following "fixes" the issue:

```js
window.Ionic = {
    config: {
      scrollAssist: false
    }
  }
```

**Comment by sean-perkins** [comment]
Ok I have no clue how my previous "fix" was working, because when running again I ran into the same behavior you noted. It was caused by the scroll assist. 

Added a workaround that increases the height of the cloned input (textarea) to the height of the container (`100%`) when used with auto grow. 

Confirmed on iOS 13 simulator that this resolves the white box rendering.

File: core/src/components/textarea/textarea.scss
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:163]: I was also able to reproduce this issue on iOS 15, so it may not be iOS 15-specific.

File: core/src/components/textarea/test/basic/index.html
Comment (liamdebeasi) [inline core/src/components/textarea/test/basic/index.html:75]: Why was this disabled?

**Review by liamdebeasi — APPROVED** [review]
Good to go when the comments below are addressed. Great job!

File: core/src/components/textarea/test/basic/index.html
Comment (sean-perkins) [reply core/src/components/textarea/test/basic/index.html:75]: When migrating the Puppeteer tests to Playwright, I ran into issues with consistently capturing the screenshots because of auto-grow. We discussed disabling those auto grow screenshots until this PR was to be merged. 

When Amanda reviewed this PR, she called out that this test is effectively a duplicate of behavior captured in the other auto grow tests, so we elected for removing it entirely from the basic tests.

**Pull request merged**
